kitsune/wiki/views.py

```diff
@@ -455,7 +455,89 @@ def steal_lock(request, document_slug, revision_id=None):
 @require_http_methods(["GET", "POST"])
 @login_required
 def edit_document(request, document_slug, revision_id=None):
-    """Create a new revision of a wiki document, or edit document metadata."""
+    """Create a new revision of a wiki document"""
+    user = request.user
+
+    doc = get_visible_document_or_404(
+        user,
+        locale=request.LANGUAGE_CODE,
+        slug=document_slug,
+        look_for_translation_via_parent=True,
+        return_parent_if_no_translation=True,
+    )
+
+    if doc.locale != request.LANGUAGE_CODE:
+        # We've fallen back to the parent, since no visible translation existed.
+        url = reverse("wiki.translate", locale=request.LANGUAGE_CODE, args=[document_slug])
+        return HttpResponseRedirect(url)
+
+    # If this document has a parent, then the edit is handled by the
+    # translate view. Pass it on.
+    if doc.parent:
+        return translate(request, doc.parent.slug, revision_id)
+    if revision_id:
+        rev = get_object_or_404(Revision, pk=revision_id, document=doc)
+    else:
+        rev = doc.current_revision or doc.revisions.order_by("-created", "-id")[0]
+
+    rev_form = None
+
+    if request.method == "GET":
+        if doc.allows(user, "create_revision"):
+            rev_form = RevisionForm(instance=rev, initial={"based_on": rev.id, "comment": ""})
+            show_revision_warning = _show_revision_warning(doc, rev)
+            locked, locked_by = _document_lock(doc.id, user.username)
+
+            return render(
+                request,
+                "wiki/edit.html",
+                {
+                    "revision_form": rev_form,
+                    "document": doc,
+                    "show_revision_warning": show_revision_warning,
+                    "locked": locked,
+                    "locked_by": locked_by,
+                },
+            )
+        # You can't do anything on this page, so get lost.
+        raise PermissionDenied
+
+    else:  # POST
+        _document_lock_clear(doc.id, user.username)
+
+        if doc.allows(user, "create_revision"):
+            rev_form = RevisionForm(request.POST)
+            rev_form.instance.document = doc  # for rev_form.clean()
+            if rev_form.is_valid():
+                _save_rev_and_notify(rev_form, user, doc, base_rev=rev)
+                if "notify-future-changes" in request.POST:
+                    EditDocumentEvent.notify(request.user, doc)
+                return HttpResponseRedirect(
+                    reverse("wiki.document_revisions", args=[document_slug])
+                )
+        else:
+            raise PermissionDenied
```

Comment:
What do you think about the following as a way to reduce nesting? I'm also assuming that the `doc.allows(user, "create_revision")` check is done as recommended above.

```suggestion
    if request.method == "GET":
        rev_form = RevisionForm(instance=rev, initial={"based_on": rev.id, "comment": ""})
        show_revision_warning = _show_revision_warning(doc, rev)
        locked, locked_by = _document_lock(doc.id, user.username)
        return render(
            request,
            "wiki/edit.html",
            {
                "revision_form": rev_form,
                "document": doc,
                "show_revision_warning": show_revision_warning,
                "locked": locked,
                "locked_by": locked_by,
            },
        )

    _document_lock_clear(doc.id, user.username)
    rev_form = RevisionForm(request.POST)
    rev_form.instance.document = doc  # for rev_form.clean()
    if rev_form.is_valid():
        _save_rev_and_notify(rev_form, user, doc, base_rev=rev)
        if "notify-future-changes" in request.POST:
            EditDocumentEvent.notify(request.user, doc)
        return HttpResponseRedirect(
            reverse("wiki.document_revisions", args=[document_slug])
        )
```

Reply:
Nice - and agreed.